File: cpp/base/src/main/vsdk/toolkit/common/logging/Logger.cpp
#include <cstdio>
#include <cstdlib>

#include "java/lang/String.h"
#include "vsdk/toolkit/common/VSDKFatalException.h"
#include "vsdk/toolkit/common/logging/Logger.h"
bool Logger::withSystemExit_ = true;
bool Logger::withFatalExceptions_ = true;

void Logger::setWithSystemExit(bool flag) { withSystemExit_ = flag; }
void Logger::setWithFatalExceptions(bool flag) { withFatalExceptions_ = flag; }

void Logger::processFatalError(const java::String& method, const java::String& message, const std::exception* cause)
{
    if ( withSystemExit_ ) {
        std::exit(1);
    }
    if ( !withFatalExceptions_ ) {
        return;
    }
    java::String m = method.empty() ? "VSDK fatal error" : ("VSDK fatal error at " + method);
    if ( !message.empty() ) {
        m += ": " + message;
    }
    if ( cause != nullptr ) {
        m += " | cause: ";
        m += cause->what();
    }
    throw VSDKFatalException(m);
}

void Logger::reportMessage(const java::String& className, int level, const java::String& method, const java::String& message)
{
    fprintf(stderr, "[VSDK][%s] %s: %s\n", className.c_str(), method.c_str(), message.c_str());
    if ( level == FATAL_ERROR ) {
        processFatalError(method, message, nullptr);
    }
}

void Logger::reportMessageWithException(const java::String& className, int level, const java::String& method, const java::String& message, const std::exception* cause)
{
    if ( cause != nullptr ) {
        fprintf(stderr, "[VSDK][%s] %s: %s | exception: %s\n", className.c_str(), method.c_str(), message.c_str(), cause->what());
    } else {
        fprintf(stderr, "[VSDK][%s] %s: %s\n", className.c_str(), method.c_str(), message.c_str());
    }
    if ( level == FATAL_ERROR ) {
        processFatalError(method, message, cause);
    }
}

File: typescript/base/src/main/vsdk/toolkit/common/logging/Logger.ts
import { Throwable } from "../../../../java/lang/Throwable.js";
import { VSDKFatalException } from "../VSDKFatalException.js";
import {
    getWithFatalExceptions,
    getWithSystemExit,
    setWithFatalExceptions,
    setWithSystemExit,
} from "../VSDKRuntimeSettings.js";

/** Platform-neutral Java Logger port. Browser runtimes report through console.error. */
export class Logger {
    public static readonly WARNING = 1;
    public static readonly ERROR = 2;
    public static readonly FATAL_ERROR = 3;
    public static reportMessageWithException(
        origin: unknown,
        level: number,
        method: string,
        message: string,
        error: Error | null,
    ): void {
        const report =
            Logger.header(origin, method, `Vitral exception message:\n${message}`) +
            (error === null
                ? " - Java exception is null! No detailed information about error.\n"
                : ` - Java exception class:\n${error.name}\n - Java exception message:\n${error.message}\n${error.stack ?? ""}\n`) +
            Logger.footer(level);
        console.error(report);
        if (error === null) console.error("Given exception is null! not reporting details!");
        else console.error(error.message, error.stack ?? "");
        Logger.fatal(level, method, message, error);
    }
    public static reportMessage(origin: unknown, level: number, method: string, message: string): void {
        console.error(Logger.header(origin, method, `Exception message:\n${message}`) + Logger.footer(level));
        Logger.fatal(level, method, message, null);
    }
    public static setWithSystemExit(flag: boolean): void {
        setWithSystemExit(flag);
    }
    public static setWithFatalExceptions(flag: boolean): void {
        setWithFatalExceptions(flag);
    }
    private static header(origin: unknown, method: string, detail: string): string {
        const type =
            origin === null || origin === undefined
                ? " - An exception has been thrown from a static context\n"
                : ` - An exception has been thrown in the "${(origin as { constructor?: { name?: string } }).constructor?.name ?? typeof origin}" class\n`;
        return `===========================================================================\n= VSDK Exception report                                                   =\n${type} - Exception located at method ${method}\n - ${detail}\n`;
    }
    private static footer(level: number): string {
        return `===========================================================================\n${level === Logger.FATAL_ERROR ? "Program excecution suspended!\n" : ""}`;
    }
    private static fatal(level: number, method: string, message: string, cause: Error | null): void {
        if (level !== Logger.FATAL_ERROR) return;
        if (!getWithSystemExit() && !getWithFatalExceptions()) return;
        const prefix = method.length === 0 ? "VSDK fatal error" : `VSDK fatal error at ${method}`;
        const text = message.length === 0 ? prefix : `${prefix}: ${message}`;
        throw new VSDKFatalException(text, cause instanceof Throwable ? cause : undefined);
    }
}
